### PR TITLE
py-h5py: add 3.8.0

### PR DESCRIPTION
**Pull request by michaelkuhn**

### CODE REVIEW

**Review by takluyver — DISMISSED** [review]

**Review by tldahlgren — COMMENTED** [review]
Confirmed the sha256.

**Comment by adamjstewart** [comment]
Also being updated in #35195

**Comment by michaelkuhn** [comment]
> Also being updated in #35195

Should we update it separately or wait for #35195 (which seems to take a while)?

**Comment by michaelkuhn** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
I'm fine with updating it separately but this PR is missing some of the changes in the other PR (upper bounds on dep versions, etc.). If you want to steal them from that PR go ahead.

**Comment by michaelkuhn** [comment]
I've pushed a new version that should now also include the correct upper bounds. I've left the setuptools@61: dependency commented out since that seems to cause problems for @kwryankrattiger.

**Comment by kwryankrattiger** [comment]
I am okay with it being separate, I wasn't too thrilled about updating it in #35195  since it is outside of the scope of the changes in there.

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-h5py/package.py]: I would uncomment this since it's an actual requirement. If it causes conflicts with other packages, then those packages will simply use an older versions of h5py. @alalazo and others are working on allowing the same package to appear in the DAG multiple times.

Comment (takluyver) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: Yup, building with an older setuptools can certainly cause issues - we discovered this recently when a bug in pip meant it used incorrect build dependencies in some situations: https://github.com/h5py/h5py/issues/2240

Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: well the reason was to placate me so the data-vis-sdk and e4s stacks can build with vol adaptors...but as of right now that is lower priority until May, do you have any ideas when duplicate build deps will be allowed?

@alalazo

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: But data-vis-sdk and e4s stacks can build with vol adaptors... using h5py 3.7.0. Or is that a feature only present in newer hdf5 supported in h5py 3.8.0? 

Either way, I don't think it's possible to support what you want until Spack can resolve multiple build deps in the DAG at the same time. As @takluyver mentioned, the 61+ requirement isn't just cosmetic, the package does not reliably build with old setuptools.

Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: h5py 3.7.0 does not support HDF5@1.13.2: which is a hard requirement for the VOLS, they don't exist before that. And the SDK heavily prefers the HDF5 1.14.0 since that is the actual release version, the odd versions are experimental.

Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: Is there a reason for packages to _not_ use a newer setuptools? I am not a heavy python user, I usually just use the latest pip gives me.

Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: I've pushed a new version that now also includes the setuptools@61: dependency.

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-h5py/package.py]:
```suggestion
    depends_on("py-cython@0.29:0", type=("build"), when="@3:")
    depends_on("py-cython@0.29.14:0", type=("build"), when="@3:3.7 ^python@3.8")
    depends_on("py-cython@0.29.15:0", type=("build"), when="@3:3.7 ^python@3.9:")
```
This constraint no longer applies in  3.8.0

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-h5py/package.py:63]: Numpy requirements also need to be loosened, not just in 3.8.0 but in older versions too

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: > Is there a reason for packages to not use a newer setuptools?

Some packages use features added in newer versions of setuptools, while other packages use features that were removed in newer versions of setuptools (e.g., 2to3). You may be able to reach out to the developers of the latter packages and ask for support for newer setuptools. Sometimes the documented dependency is incorrect and newer versions are actually supported.

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: Done

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/py-h5py/package.py:63]: Looks like they applied until 3.5. I've included the necessary changes.

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-h5py/package.py:70]: mpi4py dep needs to be updated too:
```python
    RUN_REQUIRES.append('mpi4py >=3.0.2')                                                
    SETUP_REQUIRES.append("mpi4py ==3.0.2; python_version<'3.8'")                        
    SETUP_REQUIRES.append("mpi4py ==3.0.3; python_version=='3.8.*'")                     
    SETUP_REQUIRES.append("mpi4py ==3.1.0; python_version=='3.9.*' or python_version=='3.10.*'")
    SETUP_REQUIRES.append("mpi4py ==3.1.4; python_version>='3.11'")    
```
Honestly, this stuff is kinda annoying to track. Not sure why things need to be pinned to a specific version for each Python release. Should we just ignore this and say `mpi4py@3.0.2:` for all versions? Curious what @bryanherman @takluyver think.

**Comment by takluyver** [comment]
The pinning is kind of a workaround - we aim to ensure it builds with an
older version, and then it will be compatible with newer versions at
runtime. The pinned versions are roughly the first mpi4py versions that
will conveniently work for each python version. I imagine that Spack has
better ways to express this stuff.

Needless to say, our own package metadata is a compromise between a pure
description and making things work for python tools which use that metadata
directly.

**Comment by adamjstewart** [comment]
In that case I'll let @michaelkuhn decide what to do. It sounds like `py-mpi4py@3.0.2:` should be sufficient. We don't currently use any pip/build flags to ensure that dependencies are correct, but we might in the future. If we do, I can always change this to be technically correct later.

> The pinned versions are roughly the first mpi4py versions that will conveniently work for each python version.

I'm guessing this means the first version where wheels are available, not the first version that supports that Python version? If it was the latter I would add these to the py-mpi4py package so we can remove old versions when we deprecate Python versions.

**Comment by takluyver** [comment]
> I'm guessing this means the first version where wheels are available, not the first version that supports that Python version?

That would be my guess too, though I don't know that we've applied that rule completely consistently.

**Comment by michaelkuhn** [comment]
Sorry for dropping the ball on this. I've just pushed a new version that constrains `mpi4py` to `@3.0.2:` for all 3.x versions.

File: var/spack/repos/builtin/packages/py-h5py/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-h5py/package.py]: Is this still required?

Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/py-h5py/package.py]: Looks like it was actually removed in 3.7.0 (see https://github.com/h5py/h5py/commit/d3980d3931eaa6e75066334b3d293f679d34037a), fixed.

**Review by adamjstewart — APPROVED** [review]
Looks fine to me as long as the other 2 maintainers don't see any issues.

**Comment by takluyver** [comment]
LGTM.

The precisely versioned dependencies on numpy and Cython could *probably* be simplified, like with mpi4py (discussed above). They're similar cases where we've found compromises between a pure theoretical description of dependencies, and what practically works for Python-native tooling. But if they're not causing an issue, you may just want to leave them as they are and not risk breaking anything.

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Comment by kwryankrattiger** [comment]
if no objections in the next couple hours, I am going to merge this today.

**Pull request merged**
